File: crates/chidori-js/benchmarks/workloads/mutual_recursion.js
// Function-scoped recursion in both shapes the family tiers resolve:
// MUTUAL recursion through captured bindings (isEven/isOdd — the resolver
// pins whatever compatible closure each callee-position cell holds, so the
// pair becomes a two-member family: windowed on the interpreter, two
// mutually-calling native functions under the opt-in JIT) and SELF-recursion
// through a `const` binding (gcd). Deterministic (no RNG) so every runtime
// computes the same checksum.
(function () {
  function isEven(n) {
    return n === 0 ? true : isOdd(n - 1);
  }
  function isOdd(n) {
    return n === 0 ? false : isEven(n - 1);
  }
  const gcd = (a, b) => (b === 0 ? a : gcd(b, a % b));

  const N = 20_000;
  let checksum = 0;
  for (let i = 0; i < N; i++) {
    if (isEven(i % 300)) checksum++;
    checksum += gcd(i + 123456, 991);
  }
  console.log("RESULT=" + checksum);
})();

File: crates/chidori-js/benchmarks/workloads/mutual_recursion.py
# Python twin of mutual_recursion.js — mutual recursion between two globals,
# boolean returns, and self-recursion through a local binding (gcd). Max
# depth is ~300 (is_even(299)), comfortably inside CPython's default 1000
# recursion limit.
def is_even(n):
    return True if n == 0 else is_odd(n - 1)


def is_odd(n):
    return False if n == 0 else is_even(n - 1)


gcd = lambda a, b: a if b == 0 else gcd(b, a % b)  # noqa: E731 — mirrors the JS arrow

N = 20_000
checksum = 0
for i in range(N):
    if is_even(i % 300):
        checksum += 1
    checksum += gcd(i + 123456, 991)
print("RESULT=" + str(checksum))
